pkg/cluster/docker_for_mac.go

```diff
@@ -0,0 +1,199 @@
+	if err != nil {
+		return errors.Wrap(err, "writing d4m settings")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return fmt.Errorf("writing d4m settings: status code %d", resp.StatusCode)
+	}
+	return nil
+}
+
+func (c DockerForMacClient) settings(ctx context.Context) (map[string]interface{}, error) {
+	klog.V(7).Infof("GET %s /settings\n", c.socketPath)
+	req, err := http.NewRequest("GET", "http://localhost/settings", nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading d4m settings")
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading d4m settings")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("reading d4m settings: status code %d", resp.StatusCode)
+	}
+
+	settings := make(map[string]interface{})
+	err = json.NewDecoder(resp.Body).Decode(&settings)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading d4m settings")
+	}
+	klog.V(8).Infof("Response body: %+v\n", settings)
+	return settings, nil
+}
+
+func (c DockerForMacClient) ensureK8sEnabled(settings map[string]interface{}) (changed bool, err error) {
+	vmSetting, ok := settings["vm"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing vm setting")
+	}
+
+	k8sSetting, ok := vmSetting["kubernetes"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing kubernetes setting")
+	}
+
+	enabledSetting, ok := k8sSetting["enabled"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing kubernetes enabled setting")
+	}
+
+	isEnabled, ok := enabledSetting["value"].(bool)
+	if !ok {
+		return false, fmt.Errorf("missing kubernetes enabled setting")
+	}
+
+	if isEnabled {
+		return false, nil
+	}
+	enabledSetting["value"] = true
+	return true, nil
+}
+
+func (c DockerForMacClient) ensureMinCPU(settings map[string]interface{}, desired int) (changed bool, err error) {
+	vmSetting, ok := settings["vm"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing vm setting")
+	}
+
+	resourcesSetting, ok := vmSetting["resources"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing resources setting")
+	}
+
+	cpusSetting, ok := resourcesSetting["cpus"].(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("missing resources cpus setting")
+	}
+
+	value, ok := cpusSetting["value"].(float64)
+	if !ok {
+		return false, fmt.Errorf("missing resources cpu values")
+	}
+	max, ok := cpusSetting["max"].(float64)
+	if !ok {
+		return false, fmt.Errorf("missing resources cpu max")
+	}
+
+	if desired > int(max) {
+		return false, fmt.Errorf("desired cpus (%d) greater than max allowed (%d)", desired, int(max))
+	}
+
+	if desired <= int(value) {
+		return false, nil
+	}
+
+	cpusSetting["value"] = desired
+	return true, nil
+}
+
+func (c DockerForMacClient) settingsForWrite(settings interface{}) interface{} {
```

Comment:
struggling to figure out what's going on here in each of the cases below--comments? link to spec?

Reply:
probably the tests are the best thing to look at

Reply:
I mean yes, that's what I ended up doing. that was a request for comments so that the next person to look at this code doesn't have to go spelunking in the tests to figure out what's happening.